File: py/woa.py
#!usr/bin/python

# Reference : https://en.wikiversity.org/wiki/Whale_Optimization_Algorithm

import numpy as np
import time
import sys

np.random.seed(123)

solution = {"best" : 0.,
            "walk" : [],
            "optimizer" : "",
            "objfname"  : "",
            "start_time" : 0.,
            "end_time"   : 0.,
            "execution_time" : 0.,
            "dimension" : 0.,
            "population" : 0.,
            "max_iters" : 0.
        }


def woa(objfunc,
        lower_bound,
        upper_bound,
        dim,          # Number of dimensions
        n_population, # Population size
        max_iters,    # Number of generations
        b   = 1.     #
        ):
  # Initializing arrays
  walk = np.empty(shape=(max_iters,), dtype=float)
  pos = np.random.uniform(low=lower_bound,
                          high=upper_bound,
                          size=(dim, n_population))
  # a decreases linearly fron 2 to 0 in Eq. (2.3)
  at = np.linspace(2, 0, num=max_iters)
  # a2 linearly decreases from -1 to -2 to calculate t in Eq. (3.12)
  a2t = np.linspace(-1, -2, num=max_iters)
  leader_score = np.inf
  leader_pos = np.zeros(shape=(1, dim), dtype=float)

  print ("WOA is optimizing \"" + objfunc.__name__ + "\"")
  solution["optimizer"]  = "WOA"
  solution["dimension"]  = dim,
  solution["population"] = n_population
  solution["max_iters"]  = max_iters
  solution["objfname"]   = objfunc.__name__
  solution["start_time"] = time.time()

  # main loop
  for (t, a), a2 in zip(enumerate(at), a2t):
    # Return back the search agents that go beyond the boundaries of the search space
    pos = np.clip(pos, lower_bound, upper_bound)
    # Calculate objective function for each search agent
    fitness = np.apply_along_axis(objfunc, 0, pos)

    # Update the leader
    idx = np.argmin(fitness)
    if fitness[idx] < leader_score:
      leader_score = fitness[idx]
      leader_pos   = np.array(pos[:, idx], ndmin=2).T

    r1 = np.random.uniform(low=0., high=1., size=(n_population,))
    r2 = np.random.uniform(low=0., high=1., size=(n_population,))

    A = 2. * a * r1 - a
    C = 2. * r2
    p = np.random.uniform(low=0., high=1., size=(n_population,)) < .5

    A_condition = abs(A) >= 1.

    idx = np.logical_and(p, A_condition)
    if sum(idx):
      l_idx = np.floor(n_population *
                       np.random.uniform(low=0.,
                                         high=1.,
                                         size=(n_population, ))
                       ).astype(int)
      pos[:, l_idx] = pos[:, l_idx] - A[l_idx] * abs(C * pos[:, l_idx] - pos[:, l_idx])

    idx = np.logical_and(p, ~A_condition)
    pos[:, idx] = leader_pos - A[idx] * abs((C * leader_pos)[:, idx] - pos[:, idx])

    p = ~p
    l = (a2 - 1.) * np.random.uniform(low=0., high=1., size=(n_population,)) + 1.
    l = l[p]
    dist2leader = abs(leader_pos - pos[:, p])
    pos[:, p] = dist2leader * np.exp(b * l) * np.cos(2. * l * np.pi) + leader_pos


    # Update convergence curve
    walk[t] = leader_score
    sys.stdout.write('\r')
    sys.stdout.write("It %-5d: [%-25s] %.3f %.3f sec"
                     %(t,
                       '=' * int(t / 20),
                       leader_score,
                       time.time() - solution["start_time"]))
  sys.stdout.write('\n')
  solution["end_time"] = time.time()
  solution["run_time"] = solution["end_time"] - solution["start_time"]
  solution["walk"]     = walk
  solution["best"]     = leader_score


if __name__ == "__main__":
  # F10
  score_func = lambda x: -20. * np.exp(-.2 * np.sqrt(sum(x*x) / len(x)))  \
                         - np.exp(sum(np.cos(2. * np.pi * x)) / len(x))   \
                         + 22.718281828459045

  n_population = 50
  max_iters = 500
  lower_bound = -32
  upper_bound = 32
  dim = 30

  woa(score_func,
      lower_bound,
      upper_bound,
      dim,
      n_population,
      max_iters)

File: cpp/src/woa.cpp
#include <walkers.h>

namespace walker
{
  template<typename Func>
  auto woa(Func objfunc,
           const float &lower_bound,
           const float &upper_bound,
           const int &dim,
           const int &n_population,
           const int &max_iters,
           float b = 1.f,
           std::size_t seed = 0,
           int verbose = 1,
           int nth = 4)
  {
    typedef std::pair<int, float> best_idx;
    int iteration = 0;
    best_idx leader;
    float fitness;

    std::unique_ptr<float*, std::function<void(float**)>> positions(new float*[n_population](),
                                                                    [&](float** x)
                                                                    {
                                                                      std::for_each(x, x + dim, std::default_delete<float[]>());
                                                                      delete[] x;
                                                                    });

    solution s(n_population, max_iters, "WOA");

    std::mt19937 engine(seed);
    std::uniform_real_distribution<float> bound_rng(lower_bound, upper_bound);
    std::uniform_real_distribution<float> rng(0.f, 1.f);

    // Initialize timer for the experiment
    auto start_time = std::chrono::high_resolution_clock::now();

#ifdef VERBOSE
    std::cout << "WOA is optimizing..." << std::endl;
#endif

#ifdef _OPENMP
#pragma omp declare reduction (minPair : best_idx : omp_out = omp_in.second < omp_out.second ? omp_in : omp_out) initializer(omp_priv = omp_orig)
#pragma omp parallel num_threads(nth) private(fitness)
  {
#endif

#ifdef _OPENMP
#pragma omp for
    for (int i = 0; i < n_population; ++i) (positions.get())[i] = new float[dim];
#else
    std::generate_n(positions.get(), n_population, [](){return new float[dim];});
#endif

    // initialize the positions/solutions
#ifdef _OPENMP
#pragma omp for collapse(2)
#endif
    for (int i = 0; i < n_population; ++i)
      for (int j = 0; j < dim; ++j)
        (positions.get())[i][j] = bound_rng(engine);

    // main loop
    while(iteration < max_iters)
    {
      leader.second = inf;
      leader.first  = 0;
#ifdef _OPENMP
#pragma omp for reduction(minPair : leader)
#endif
      for (int i = 0; i < n_population; ++i)
      {
        // compute objective function for each search agent
        fitness = objfunc((positions.get())[i]);
        // update the leader
        leader.first  = fitness < leader.second ? i       : leader.first; // update alpha
        leader.second = fitness < leader.second ? fitness : leader.second;
      }

      float a  = 2.f - iteration * (2.f / max_iters); // a decreases linearly from 2 to 0 in Eq. (2.3)
      float a2 = 1.f + iteration * (-1.f / max_iters);// a2 linearly decreases from -1 to -2 to calculate t in Eq. (3.12)

      // update the position of search agents
#ifdef _OPENMP
#pragma omp for
#endif
      for (int i = 0; i < n_population; ++i)
      {
        float r1 = rng(engine);
        float r2 = rng(engine);

        float A    = 2.f * a * r1 - a; // Eq. (2.3) in the paper
        float C    = 2.f * r2;         // Eq. (2.4) in the paper
        float l    = (a2 - 1.f) * rng(engine) + 1.f; // parameters in Eq. (2.5)
        bool p     = static_cast<bool>(rng(engine) > .5f);      // Eq. (2.6)
        bool a_abs = static_cast<bool>(std::fabs(A) >= 1.f);

        for (int j = 0; j < dim; ++j)
        {
          //(positions.get())[i][j] = (p && a_abs)  ? (positions.get())[/*MISS*/][j] - A * std::fabs(C * (positions.get())[/*MISS*/][j] - (positions.get())[i][j])                         :
          //                          (p && !a_abs) ? (positions.get())[leader.first][j] - A * std::fabs(C * (positions.get())[leader.first][j] - (positions.get())[i][j]) :
          //                          std::fabs((positions.get())[leader.first][j] - (positions.get())[i][j]) * std::exp(b * l) * std::cos(l * 2.f * M_PI) + (positions.get())[leader.first][j];
          (positions.get())[i][j] = ((positions.get())[i][j] < lower_bound) ? lower_bound :
                                    ((positions.get())[i][j] > upper_bound) ? upper_bound :
                                     (positions.get())[i][j];
        }
      }

      s.walk[iteration] = (positions.get())[leader.first];

#ifdef _OPENMP
#pragma omp single
#endif
      ++iteration;
    } // end while

#ifdef _OPENMP
  } // end parallel section
#endif

    auto end_time = std::chrono::high_resolution_clock::now();
    s.execution_time = std::chrono::duration_cast<std::chrono::seconds>(end_time - start_time).count();
    s.best = leader.second;

    return s;
  }

}

